boa_engine/src/value/mod.rs

```diff
@@ -88,184 +132,817 @@ impl JsValue {
+        debug_assert_eq!(value.tag(), ValueTag::BigInt);
+        value
+    }
+}
+
+/// This abstracts over every pointer type boxed inside `NaN` values.
+///
+/// # Safety
+///
+/// Non-exhaustive list of situations that could cause undefined behaviour:
+/// - Returning an invalid `*mut ()`.
+/// - Returning a `ManuallyDrop<Self>` that doesn't correspond with the provided
+/// `ptr`.
+/// - Dropping `ty` before returning its pointer.
+pub(crate) unsafe trait PointerType {
+    unsafe fn from_void_ptr(ptr: *mut ()) -> ManuallyDrop<Self>;
+
+    unsafe fn into_void_ptr(ty: ManuallyDrop<Self>) -> *mut ();
+}
+
+impl Drop for JsValue {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            match self.tag() {
+                ValueTag::Object => {
+                    ManuallyDrop::into_inner(JsObject::from_void_ptr(self.as_pointer()));
+                }
+                ValueTag::String => {
+                    ManuallyDrop::into_inner(JsString::from_void_ptr(self.as_pointer()));
+                }
+                ValueTag::Symbol => {
+                    ManuallyDrop::into_inner(JsSymbol::from_void_ptr(self.as_pointer()));
+                }
+                ValueTag::BigInt => {
+                    ManuallyDrop::into_inner(JsBigInt::from_void_ptr(self.as_pointer()));
+                }
+                _ => {}
+            }
         }
     }
+}
 
-    /// Returns true if the value is a number.
+impl Clone for JsValue {
     #[inline]
-    pub fn is_number(&self) -> bool {
-        matches!(self, Self::Rational(_) | Self::Integer(_))
+    fn clone(&self) -> Self {
+        unsafe {
+            match self.tag() {
+                ValueTag::Object => Self::new(self.as_object_unchecked().clone()),
+                ValueTag::String => Self::new(self.as_string_unchecked().clone()),
+                ValueTag::Symbol => Self::new(self.as_symbol_unchecked().clone()),
+                ValueTag::BigInt => Self::new(self.as_bigint_unchecked().clone()),
+                _ => Self {
+                    value: Cell::new(self.value.get()),
+                },
+            }
+        }
     }
+}
+
+impl Finalize for JsValue {}
 
+unsafe impl Trace for JsValue {
     #[inline]
-    pub fn as_number(&self) -> Option<f64> {
-        match *self {
-            Self::Integer(integer) => Some(integer.into()),
-            Self::Rational(rational) => Some(rational),
-            _ => None,
+    unsafe fn trace(&self) {
+        if let Some(o) = self.as_object() {
+            // SAFETY: `self.as_object()` must always return a valid `JsObject
+            unsafe { o.trace() }
         }
     }
 
-    /// Returns true if the value is a string.
     #[inline]
-    pub fn is_string(&self) -> bool {
-        matches!(self, Self::String(_))
+    unsafe fn root(&self) {
+        if let Some(o) = self.as_object() {
+            // SAFETY: `self.as_object()` must always return a valid `JsObject
+            unsafe { o.root() }
+        }
```

Comment:
```suggestion
    unsafe fn root(&self) {
        if let Some(o) = self.as_object() {
            // SAFETY: `self.as_object()` must always return a valid `JsObject
            unsafe { o.root() }
        }
```
This will not root the object contained in `JsValue`, this will only root local object `o` (it's least significat bit is set, and does not update self)

This is why self is `Cell<u64>` and was set back after `root` and `unroot`.

Reply:
AH RIGHT!
To be honest I've been thinking counfused for like 20 full minutes, but I forgot that `JsValue` now is a `Cell`, thanks.